File: Clojure/problem027.clj
#! /usr/bin/env clojure
; Problem 27 - Project Euler
; http://projecteuler.net/index.php?section=problems&id=027

(require '[mathtools :refer [primes ppp]])

(def primes100m (primes 1000000))

(defn prime-counter [a b]
  (loop [n 0 c 0]
    (let [p (+ (* n n) (* a n) b)]
      (if (< (.indexOf primes100m (Math/abs p)) 0)
        (list a b c)
        (recur (inc n) (inc c))))))

(->> (for [a (range -999 1000)
           b (primes 1000)
           :let  [z (prime-counter a b)]
           :when (> (nth z 2) 2)]
       z)
     (apply max-key #(nth % 2))
     (ppp)
     (take 2)
     (reduce *)
     (println))
    

File: Ruby/problem027.rb
#! /usr/bin/env ruby
=begin
Problem 27 - Project Euler
http://projecteuler.net/index.php?section=problems&id=27
=end
require 'prime'
def count_primes(a,b)
    n = 0
    while (n**2 + a*n + b).prime?
        n += 1
    end
    n
end

puts (-999..999).flat_map {|a|
    (2..1000).select(&:prime?).map {|b|
        [a, b, count_primes(a,b)]
    }
}.max_by {|x| x[2]}.take(2).inject(:*)
